bpf: nat: support ICMPV6_DEST_UNREACH in egress path

julianwiedmann opened the pull request:
Add the missing bits in the IPv6 SNAT engine, so that `ICMPV6_DEST_UNREACH` packets (which we generate for a NodePort service without backends) can exit the node.

julianwiedmann committed:
bpf: nat: support ICMPV6_DEST_UNREACH in egress path

Add the minimal bits to handle ICMPv6 error messages in the IPv6 NAT
egress path. This is basically a copy of the corresponding ICMPv4 code.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: test: validate service-no-backend-response vs BPF Masquerading

The LB path sends an ICMP error message when the accessed service has no
backend available (see https://github.com/cilium/cilium/pull/28157).

In the case of a nodeport service, validate that this ICMP error message
can exit the node (and traverse the `to-netdev` program) when
BPF masquerading is enabled.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

dylandreimerink reviewed (approved):
This looks good to me. Though based on CI output it seems like the tests need some work.

julianwiedmann committed:
bpf: lb: refactor tail_no_service_ipv*()

Slim down the inner function __tail_no_service_ipv*() sufficiently, so
that it contains *only* the packet-rewrite logic. This allows us to easily
reuse it from test context, to build identical packets.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
bpf: lb: use redirect_self() in tail_no_service_ipv*()

Use the intended helper to be a bit more self-documenting.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

dylandreimerink reviewed (approved):
:rocket:
